cpp/include/cudf/interop.hpp

```diff
@@ -412,6 +416,24 @@ std::unique_ptr<table> from_arrow_host(
   rmm::cuda_stream_view stream        = cudf::get_default_stream(),
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Create `cudf::table` from given ArrowArrayStream input
+ *
+ * @throws std::invalid_argument if input is NULL
+ *
+ * The conversion will not call release on the input ArrayArrayStream or its
+ * constituent arrays or schema.
```

Comment:
This might be related to the thread about moving (https://github.com/rapidsai/cudf/pull/15904#discussion_r1628151778), but what is the reason this method will not release the input? 

When consuming through the C Data Interface, the consumer should move the input ArrowArray (this is to ensure that the no-one else will call the release calback while you are using this data. For example, the capsule destructor will call the release callback if that is not set to NULL; moving will set that to NULL avoiding anyone else to call it). 
In this case, I am also assuming that you are copying the data here from CPU to CUDA device? If that's the case, I would assume you should also be able to release all input data at the end of this function?

Reply:
Thanks for chiming in @jorisvandenbossche! In this case, the reason boils down to the fact that this API is just looping over the chunks and calling `from_arrow` on each of them, and [that method does not release the array](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/include/cudf/interop.hpp#L360). That was added in https://github.com/rapidsai/cudf/pull/15645. Perhaps that function should also be releasing the host? It's possible that some of the release semantics that we had put into the different `from_arrow_device` overloads for view vs copies were incorrectly propagated to the host API and I didn't catch it due to my relative unfamiliarity with the protocol. I would expect that the release semantics of `from_arrow_stream` vis-a-vis releasing should be entirely inherited from `from_arrow` in this case.

This discusssion is related to the thread about moving; not to that exact code, but certainly to some of my confusion on that front. You are right that we are copying the data from host to device in this function. In the thread that you linked I illustrated a case that would require that the underlying host allocation not be freed even after this function completes: when the source of the stream is a pyarrow array that still needs to be valid after the stream is destroyed.

I think the key point here is to distinguish releasing an ArrowArray from actually freeing the underlying buffer, so let me sketch out my understanding here and verify that it makes sense. When it is initially produced, an ArrowArray may be the sole owner of data, or it may not. The release semantics of ArrowArray are designed to support the first case up front: if the ArrowArray owns its data, then the release callback can only be called once since it is directly responsible for freeing the data, so we have to ensure that it is only called once to avoid a double free. To support the second case where the initially produced ArrowArray is already sharing data (presumably with its source, such as a pyarrow array), then the release callback and the private_data member must be designed to handle that second level of sharing so that releasing the ArrowArray does not actually free the data unless the other shared owners are also ready to release it (e.g. via an underlying shared_ptr). Is that the right way to think about it?

If so, are there any scenarios where it is valid to copy, rather than move, an ArrowArray? IOW, can you ever have two ArrowArrays referencing the same data and expect both to independently keep the data alive for the duration of their scope? I ask this noting that the spec [explicitly discusses moving arrays](https://arrow.apache.org/docs/format/CDataInterface.html#moving-an-array) but not copying them.

Reply:
> In the thread that you linked I illustrated a case that would require that the underlying host allocation not be freed even after this function completes: when the source of the stream is a pyarrow array that still needs to be valid after the stream is destroyed.

Which case are you exactly referring to? You mean the one of a pyarrow.Table producing a stream that you consume and copy to device memory? 
It's related to my previous comment, but you as a consumer don't need to worry about that in this case. Calling the release callback doesn't necessarily deallocate anything right away. It only signals to the actual owner of the data (the pyarrow.Table) that you no longer need the data. And if at that point the pyarrow.Table is also not needed for anything else (no other reference to keep it alive), then it will be actually deallocated. But if not, it will just stay alive.

> I think the key point here is to distinguish releasing an ArrowArray from actually freeing the underlying buffer, so let me sketch out my understanding here and verify that it makes sense. ... Is that the right way to think about it?

That's generally correct, yes. But I don't know if it is necessary to distinguish those two cases of being the sole owner or not. In practice, I think in most implementations the ArrowArray will almost never be the sole owner, because a typical implementation (maybe with nanoarrow as an exception) will have its own objects to represent Arrow data, and not directly use an ArrowArray struct. So at the moment you export your data to an ArrowArray, you typically let that ArrowArray point to your own object (even if that object would not be used for anything else as being kept alive by the ArrowArray). So that distinction is already a bit fluid.

But the important aspect is that in either case the release callback should always only be called once. This is ensured by the spec stating that a release callback should always set the `release` member of the struct to NULL (https://arrow.apache.org/docs/format/CDataInterface.html#release-callback-semantics-for-producers), making the ArrowArray marked as "released" (https://arrow.apache.org/docs/format/CDataInterface.html#released-structure), and avoiding that the release callback can be called a second time.

> If so, are there any scenarios where it is valid to copy, rather than move, an ArrowArray? IOW, can you ever have two ArrowArrays referencing the same data and expect both to independently keep the data alive for the duration of their scope? I ask this noting that the spec [explicitly discusses moving arrays](https://arrow.apache.org/docs/format/CDataInterface.html#moving-an-array) but not copying them.

No, so you should never copy the struct, but only move it. Otherwise you can have two structs pointing to the same data with the same release callback but that can be called from each struct, i.e. allowing to call it twice which is not allowed.

Reply:
> In this case, the reason boils down to the fact that this API is just looping over the chunks and calling `from_arrow` on each of them, and [that method does not release the array](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/include/cudf/interop.hpp#L360). That was added in #15645. Perhaps that function should also be releasing the host?

On the original topic of why this method is not calling the release callback: whether `from_arrow` is calling the release callback or not depends on you want to structure the code and where to put the responsibilities to deal with this. I think for internal usage, it can be perfectly fine to state that `from_arrow` does not release the passed ArrowArray itself, and that this is the responsibility of the caller of `from_arrow`. 

You only need to ensure that the data that you consume through the `__arrow_c_array__` protocol is released _somewhere_. Looking at the code that does this in this PR:

```
def _from_arrow_column(pyarrow_object, *, DataType data_type=None):
    if data_type is not None:
        raise ValueError("data_type may not be passed for arrays")

    schema, array = pyarrow_object.__arrow_c_array__()
    cdef ArrowSchema* c_schema = (
        <ArrowSchema*>pycapsule.PyCapsule_GetPointer(schema, "arrow_schema")
    )
    cdef ArrowArray* c_array = (
        <ArrowArray*>pycapsule.PyCapsule_GetPointer(array, "arrow_array")
    )

    cdef unique_ptr[column] c_result
    with nogil:
        c_result = move(cpp_from_arrow_column(c_schema, c_array))

    return Column.from_libcudf(move(c_result))
```

The above can actually be fine _if_ you didn't move the ArrowArray and the data is actually being copied. Because what happens in the above is that on returning from this method, the capsules (`schema`, `array`) will go out of scope, which will call their destructors, which in turn will call the release callback of the ArrowArray they hold.

But for a function like the above, there are generally two cases:

- You fully copy the data, and it is fine to release the ArrowArray upon returning the result from the method
- You don't copy the data, but the returned cudf object is a (partial) zero-copy view on the ArrowArray (which could be the case if when working with a DeviceArrowArray). At that point you need to ensure that the release callback does not get called at the end of this method, which can be achieved by 1) moving the ArrowArray struct (so the capsule destructor, which still gets called, will not call the release callback), and 2) keeping track of the ArrowArray in the cudf object (ensuring that when the cudf object goes out of scope, it will eventually call the release callback of the ArrowArray).

So I _think_ that moving+calling the release callback is not strictly needed in this PR because you always copy the data and the ArrowArray can directly be released. But it might still be cleaner / clearer to explicitly handle that on your side instead of relying on the capsule destructor.

Reply:
Yes, cudf has two classes with very clear ownership semantics, which is what complicates the interchange a bit. Let's consider the four possible `from_arrow*` APIs that we could have (using ArrowArray as a proxy for host data and ArrowDeviceArray as a proxy for device data):
1. `std::unique_ptr<column_view> from_arrow(ArrowArray *)`: This doesn't make sense because cudf can't create a (device) view onto host data.
2. `std::unique_ptr<column> from_arrow(ArrowArray *)`: We have this overload. The above discussion is around whether it should release the array or not. I argued that it shouldn't since we are always making a copy of the data anyway (we're taking host data and returning device data). @paleolimbot you think it would be better to release, is that right? I'm open to being convinced of that. Based on your suggestion, I could certainly see an argument for this accepting a `const ArrowArray *` instead to broadcast that we aren't going to release, so that seems like a good change to make if we stick with not releasing.
3. `std::unique_ptr<column_view> from_arrow(ArrowDeviceArray *)`: We have this overload. Currently it does not call the `release` callback and leaves that up to the caller. However, because there are some data types (e.g. bools) that libcudf represents differently from Arrow, there may be some data allocated as part of this conversion. As a result, the returned `unique_ptr` has a custom deleter for releasing that data. One thing we could do here is update that deleter to also call the `release` callback. In that case it would no longer be the caller's responsibility to release. Does that choice align more or less with your expectations?
4. `std::unique_ptr<column> from_arrow(ArrowDeviceArray *)`: This doesn't make sense because cudf can't take sole ownership of the data and that is basically a requirement of the object. The data pointer is hidden under a couple levels of RAII object abstraction, so there's no magic with custom deleters that we could do to avoid freeing that pointer out from under a caller that expects to be handing over shared ownership of the data (i.e. we can't just call the `release` callback, we will also always `free` the underlying allocation).

Reply:
Yeah we could use a `column_view`. It's certainly an abuse since it's not really the object itself that has the necessary semantics but rather the `unique_ptr` we're wrapping it with, but that's OK. We're are already doing it for the required allocations in (3) so I could see an argument for generalizing that to also handle the release.

Shared ownership is certainly convenient, but it has drawbacks at scale. When resource management is paramount (as it is with GPU memory) you really want ownership semantics to be unambiguous. With a `shared_ptr` (or a ref-counting language like Python) you never really know when an object is being deleted. Having libcudf's objects be movable but not copyable makes that much clearer. The ownership semantics don't usually restrict callers of libcudf algorithms since they all operate on views, so that you can process data managed by other frameworks too. The main assumption is that the application using libcudf can always make a better decision about when to free the data than libcudf itself. That doesn't fully align with the way you might expect to share Arrow data across library boundaries, though.

Reply:
Instead of always returning either a `column` or a `column_view`, what if you returned a wrapping object, i.e. something like an `ArrowDeviceArrayWrapper`. This class could have a method to get a `column_view` zero copy, a method to deep copy into a `column` (since you can't transfer ownership), and have a destructor that is responsible for calling the `release` callback and doing whatever other cleanup is necessary? From my perspective that at least feels easier and cleaner to use as a consumer of libcudf then to reason about calling the release callback and managing cleanup myself.

Reply:
> Instead of always returning either a column or a column_view, what if you returned a wrapping object, i.e. something like an ArrowDeviceArrayWrapper.

This is where my head went too after I wrote up my last response. Fundamentally we want a memory-sharing analog to `cudf::column` that can produce `cudf::column_view` objects. The implementation details aren't important, only the different ownership semantics, but introducing a new class would be a cleaner approach than hacking the logic into the `unique_ptr`'s deleter. The deep copying to a `cudf::column` isn't necessary since it would already be achievable with the one-liner `auto col = std::make_unique<column>(arrow_array_wrapper->view);`, but otherwise yes I think we're on the same page. It would certainly allow us to match the Arrow expected release semantics far more cleanly.

Since that is a change to the device array processing, it doesn't have any immediate bearing on this PR, so I'll make that change in a separate one. For this PR I'll just focus on getting the host and stream versions behaving in accordance with all the logic we've discussed above.

Reply:
Also, @paleolimbot over the course of all these discussions I completely forgot that when we originally reviewed the code Matt wrote we already got all the `const` qualifiers correct! We had the right ideas about what how the semantics of the arguments would translate to how the object was used, but I hadn't fully grasped the broader Arrow release semantics and got a bit muddled about the current state of things while trying to figure that out. So the main change to be made here (in a future PR) is to change what is returned from the device-side `from_arrow` APIs.

Reply:
Continuing in #16104.